include/nil/crypto3/zk/commitments/detail/polynomial/eval_storage.hpp

```diff
@@ -0,0 +1,98 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+//---------------------------------------------------------------------------//
+
+#ifndef CRYPTO3_ZK_PLACEHOLDER_EVAL_STORAGE_HPP
+#define CRYPTO3_ZK_PLACEHOLDER_EVAL_STORAGE_HPP
+
+#include <nil/crypto3/zk/transcript/fiat_shamir.hpp>
+#include <nil/crypto3/zk/commitments/type_traits.hpp>
+
+namespace nil {
+    namespace crypto3 {
+        namespace zk {
+            namespace commitments {
+                template<typename FieldType>
+                class eval_storage{
+                private:
+                    std::map<std::size_t, std::vector<std::vector<typename FieldType::value_type>>> z;
+                public:
+                    bool operator==(const eval_storage& other) const{
+                        return this->z == other.z;
+                    }
+                    eval_storage &operator=(const eval_storage& other){
+                        this->z = other.z;
+                        return *this;
+                    }
+                    std::vector<std::size_t> get_batches() const{
+                        std::vector<std::size_t> batches;
+
+                        for(auto it = z.begin(); it != z.end(); ++it){
+                            batches.push_back(it->first);
+                        }
+                        return batches;
+                    }
+                    std::map<std::size_t, std::size_t> get_batch_info() const{
+                        std::map<std::size_t, std::size_t> batch_info;
+
+                        for(auto it = z.begin(); it != z.end(); ++it){
+                            batch_info[it->first] = it->second.size();
+                        }
+                        return batch_info;
+                    }
+                    std::size_t get_batches_num() const{
+                        return z.size();
+                    }
+                    std::size_t get_batch_size(std::size_t batch_id) const{
+                        return z.at(batch_id).size();
+                    }
+                    std::size_t get_poly_points_number(std::size_t batch_id, std::size_t poly_id) const{
+                        return z.at(batch_id)[poly_id].size();
+                    }
+                    const std::vector<std::vector<typename FieldType::value_type>> &get(std::size_t batch_id) const{
+                        return z.at(batch_id);
+                    }
+                    const std::vector<typename FieldType::value_type> &get(std::size_t batch_id, std::size_t poly_id) const{
+                        return z.at(batch_id)[poly_id];
+                    }
+                    typename FieldType::value_type get(std::size_t batch_id, std::size_t poly_id, size_t point_id) const{
```

Comment:
Please make this one's return type a const reference as well, it's called a lot of times from marshalling-zk.